File: flux-framework/src/main/java/jp/bglb/bonboru/flux/middleware/Middleware.java
package jp.bglb.bonboru.flux.middleware;

import jp.bglb.bonboru.flux.action.ActionData;
import jp.bglb.bonboru.flux.action.ActionType;
import jp.bglb.bonboru.flux.store.Store;
import rx.Observable;

/**
 * Created by Tetsuya Masuda on 2016/07/01.
 */

public class Middleware<T, E extends ActionType> {

  public void before(Store<T> store) {
  }

  public Observable<ActionData<T, E>> intercept(Observable<ActionData<T, E>> observable) {
    return observable;
  }

  public ActionData<T, E> after(Store<T> store, ActionData<T, E> actionData) {
    return actionData;
  }
}

File: flux-framework-kotlin/src/main/kotlin/jp/bglb/bonboru/flux/kotlin/middleware/Middleware.kt
package jp.bglb.bonboru.flux.kotlin.middleware

import jp.bglb.bonboru.flux.action.ActionData
import jp.bglb.bonboru.flux.action.ActionType
import jp.bglb.bonboru.flux.store.Store
import rx.Observable

/**
 * Created by tetsuya on 2016/07/01.
 */
open class Middleware<T, E : ActionType>() {

  open fun before(store: Store<T>) {
  }

  open fun intercept(
      observable: Observable<ActionData<T, E>>): Observable<ActionData<T, E>> = observable

  open fun after(store: Store<T>, actionData: ActionData<T, E>): ActionData<T, E> = actionData
}
